### PR TITLE
feat: add products page

### PR DESCRIPTION
**Pull request by anik31**
## Products Page
I have added the products page. Ignore the CSS & backend db files. Focus only on the components, pages & context + reducer files.
 
## Files to check
- [ ]  src/App.js
- [ ]  src/context/product-context.js
- [ ]  src/hooks/useAsyncFetch.js
- [ ]  src/pages/home/CategoryCard.jsx
- [ ]  src/pages/products/Filter.jsx
- [ ]  src/pages/products/Products.jsx
- [ ]  src/reducer/product-reducer.js
- [ ]  src/utils/discountCalc.js
- [ ]  src/utils/filterData.js
- [ ]  src/utils/sortData.js
 
 _PS - Ignore other files._

image

### CODE REVIEW

File: src/components/cards/ProductCard.jsx [hunk 1]
```diff
@@ -1,13 +1,27 @@
+import { discountCalc } from "../../utils";
+import "./cards.css";
+
 export function ProductCard({ value }) {
-  const {outOfStock, badge, image, title, price} = value;
+  const {outOfStock, badge, image, title, price, originalPrice, rating} = value;
```
Comment (Maianki) [inline src/components/cards/ProductCard.jsx:5]: Nitpick: You can also destructure image by doing something like this ```const {outOfStock, badge, image:{src,alt}, title, price, originalPrice, rating} = value; ```

File: src/hooks/useAsyncFetch.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+            try{
+                const {data} = await axios.get(url);
+                console.log("Calling api"+url);
+                dispatch({type:dispatchType, payload: data[dispatchPayload]});
+            }catch(err){
+                console.log(err);
+            }
+        })();
+    },[]);
+}
```
Comment (Maianki) [inline src/hooks/useAsyncFetch.js:18]: Loved that you have created a custom hook for all API calls.

File: src/pages/home/CategoryCard.jsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { useNavigate } from "react-router-dom";
+import { useProducts } from "../../context/product-context";
+
+export function CategoryCard({value}){
+    const {categoryName,image} = value;
```
Comment (Maianki) [inline src/pages/home/CategoryCard.jsx:5]: Here also you can destructure the image too.

**Review by Maianki — APPROVED** [review]
Great job Aniket loved your code. It's well structured and follows consistent code practices. Has dropped a few suggestions.

File: src/context/product-context.js [hunk 4]
```diff
@@ -3,11 +3,21 @@ import { productReducer } from "../reducer/product-reducer";
...
+        includeOutOfStock: false,
+    }
+};
+
 const ProductProvider = ({children}) => {
-    const [state, dispatch] = useReducer(productReducer,{
-        products:[],
-        categories:[]
-    })
+    const [state, dispatch] = useReducer(productReducer, initialState);
```
Comment (nayyyhaa) [inline src/context/product-context.js:19]: Nice that you've defined initialState as a separate object.

File: src/hooks/useAsyncFetch.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import {useEffect} from "react";
+import axios from "axios";
+
+export function useAsyncFetch({url, dispatchType, dispatchPayload}){
+    const {dispatch}  = useProducts();
+    useEffect(()=>{
+        (async function asyncProductFetch(){
+            try{
+                const {data} = await axios.get(url);
+                console.log("Calling api"+url);
```
Comment (nayyyhaa) [inline src/hooks/useAsyncFetch.js:11]: can remove the unused console.log()s from the code.

**Review by nayyyhaa — APPROVED** [review]
image
here you can give labels for the price range for better understanding. Otherwise, nice work! Loved the use of custom hooks.

File: src/components/cards/ProductCard.jsx (hunk 1 above)
Comment (anik31) [reply src/components/cards/ProductCard.jsx:5]: Okay, will do it later.

File: src/hooks/useAsyncFetch.js (hunk 2 above)
Comment (anik31) [reply src/hooks/useAsyncFetch.js:18]: Thank you

File: src/pages/home/CategoryCard.jsx (hunk 3 above)
Comment (anik31) [reply src/pages/home/CategoryCard.jsx:5]: Okay, will do it later

File: src/context/product-context.js (hunk 4 above)
Comment (anik31) [reply src/context/product-context.js:19]: Yes, to make it more readable

File: src/hooks/useAsyncFetch.js (hunk 5 above)
Comment (anik31) [reply src/hooks/useAsyncFetch.js:11]: I have added it intentionally to keep a watch on the number of API calls being made, so will try to reduce the number of API calls if found something is redundant. After the app is complete I will remove the console.log(), for now I will keep it.

**Pull request merged**